Error if derivative mode other than "rev" is requested for `OM_DVGEOCOMP`

eytanadler opened the pull request:
## Purpose
<!--
Explain the goal of this PR, if it addresses an existing issue be sure to link to it.
Describe the big picture of your changes here, perhaps using a bullet list if multiple changes are done to accomplish a single goal.
If this PR accomplishes multiple goals, it may be best to create separate PR's for each.
-->
The MPhys component for pygeo supports only reverse mode derivatives and otherwise gives zero derivatives. However, no warning or error is thrown. This PR adds a `RuntimeError` if the derivatives are requested with a mode other than `"rev"`.

## Expected time until merged
<!--
Comment on whether or not this change is urgent and how long you expect this PR to take to review and be merged.
For example, a week would be a reasonable time frame for an average, non-urgent PR.
-->
A few days

## Type of change
<!--
What types of change is it?
Select the appropriate type(s) that describe this PR
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

## Testing
<!-- Explain the steps needed to test the new code to verify that it does indeed address the issue and produce the expected behavior. -->

## Checklist
<!-- Put an `x` in the boxes that apply. -->

- [ ] I have run `flake8` and `black` to make sure the Python code adheres to PEP-8 and is consistently formatted
- [ ] I have formatted the Fortran code with `fprettify` or C/C++ code with `clang-format` as applicable
- [ ] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

hajdik reviewed (dismissed):
Thanks for catching this, LGTM

eirikurj commented:
@eytanadler, this was discussed during the maintenance meeting last week (missed the last one), but just want to clarify, do you plan to add the forward mode in this PR? We can also just add that in a separate PR, but in that case please make an issue documenting the forward mode additions etc.

eytanadler commented:
Yes, I’ll add it here. I just haven’t got around to it yet.

eytanadler commented:
I did a first pass of forward mode derivatives for the MPhys DVGeo component. I'm not super familiar with the parallelization hacks, so @anilyil please check those and let me know if you have any concerns. Tests for this component and its derivatives are included on the mphys-tests branch, so I think it'd be best to either merge that branch into this one or this branch into that one. That way, we can ensure these derivatives work as expected before merging to main. I ran the tests locally and this seems to produce the same results as reverse mode.

A couple other things I found and changed while making this:
- The check for missing variables in `totalSensitivityProd` incorrectly finds missing variables if the variable is not in the final DVGeo. I fixed this by initializing the set with the variables we want to find rather than adding them as we go.
- The tests on the mphys-tests branch identify an incorrect derivative in the projected area constraint that is not tested elsewhere in the code. I looked into it for awhile and couldn't nail down the error. However, I did find one case where the inequality check in the derivative calculation is `if XYZ > 0 [do something] else [do something else]`, whereas in the primal it is `if XYZ < 0 [do something else] else [do something]`. I switched the derivative version around so it matches the primal. @hajdik it'd be good for us to get to the bottom of this derivative problem. Nonetheless, the fwd and rev modes have the same problem, so I don't think it's something I introduced in this code.

In the meantime, I'll add fwd mode derivative checking to the tests on the mphys-tests branch.

timryanb commented:
What's the status of this PR @eytanadler? I made my own implementation of fwd mode (PR #254), but looking at this PR in further detail you appear to have already added your own implementation. Is this working? If so, I can close my PR.

eytanadler commented:
> What's the status of this PR @eytanadler? I made my own implementation of fwd mode (PR #254), but looking at this PR in further detail you appear to have already added your own implementation. Is this working? If so, I can close my PR.

@timryanb this PR is ready to go, but the tests for it are on the [mphys-tests branch](https://github.com/mdolab/pygeo/tree/mphys-tests). @hajdik and I were planning on wrapping that up and merging with this, but then got sidetracked with other PhD work. One of the tricky things with that work was we found an incorrect derivative and spent some time digging to figure out why, but to no avail, which is one of the reasons it got held up.

As an aside, are there any details from your forward mode implementation that I missed here or that would improve this PR's implementation?

timryanb commented:
> > What's the status of this PR @eytanadler? I made my own implementation of fwd mode (PR #254), but looking at this PR in further detail you appear to have already added your own implementation. Is this working? If so, I can close my PR.
> 
> @timryanb this PR is ready to go, but the tests for it are on the [mphys-tests branch](https://github.com/mdolab/pygeo/tree/mphys-tests). @hajdik and I were planning on wrapping that up and merging with this, but then got sidetracked with other PhD work. One of the tricky things with that work was we found an incorrect derivative and spent some time digging to figure out why, but to no avail, which is one of the reasons it got held up.
> 
> As an aside, are there any details from your forward mode implementation that I missed here or that would improve this PR's implementation?

No, I think our approaches should be functionally equivalent. I tried running your branch on a test case of mine and it ran as expected, so I think I'll just close my PR.

eytanadler commented:
This is finally ready for review

lamkina commented on pygeo/mphys/mphys_dvgeo.py:
```diff
@@ -215,6 +219,9 @@ def nom_addPointSet(self, points, ptName, add_output=True, DVGeoName=None, **kwa
             self.add_output(ptName, distributed=True, val=points.flatten())
 
+        if isinstance(DVGeo, DVGeometryESP):
+            return dMax_global
```
I prefer to not use conditional returns, can we tweak this to set `dMaxGlobal` equal to `None` if the `DVGeo` type is not `DVGeometryESP`? Then just return the variable without the if-statement.

lamkina commented on pygeo/mphys/mphys_dvgeo.py:
```diff
@@ -997,34 +1040,45 @@ def compute_jacvec_product(self, inputs, d_inputs, d_outputs, mode):
+                                    # check if this dv is present
+                                    if k in d_inputs:
+                                        # do the allreduce
+                                        # TODO remove the allreduce when this is fixed in openmdao
```
I think we can remove this TODO. MPhys and OM dev teams decided that individual codes, not OM, are responsible for correctly performing the `allreduce` in the matrix-free computation.

lamkina commented on pygeo/mphys/mphys_dvgeo.py:
```diff
@@ -997,34 +1040,45 @@ def compute_jacvec_product(self, inputs, d_inputs, d_outputs, mode):
+                                        xdotg[k] = self.comm.allreduce(xdot[k], op=MPI.SUM)
+
+                                        # accumulate in the dict
+                                        # TODO
```
If this is working well I think we can clean up this comment as well. If there's actually plans to fix and move back to `totalSensitivityTransProd` then we should create and issue and reference this code in the issue. Otherwise this comment will probably get lost to the ether.

lamkina commented on tests/reg_tests/commonUtils.py:
```diff
@@ -283,3 +283,55 @@ def spanX(val, geo):
+                i_weight = (n_chord - ii - 1) / (n_chord - i_center - 1)
+
+            # get the direction vectors with unit length
+            dir_up = np.array([0.0, 1.0, 0.0])
```
Are these Cartesian direction vectors? Should they be an input? I'm not sure the +y and -y will always be considered up/down for all topologies.

Disregard if I'm misunderstanding this.

lamkina commented on tests/reg_tests/test_MPhysGeo.py:
```diff
@@ -0,0 +1,566 @@
+
+# input files for all DVGeo types
+input_path = os.path.dirname(os.path.abspath(__file__))
+outerFFD = os.path.join(input_path, "../../input_files/outerBoxFFD.xyz")
```
Should change `"../../input_files/etc"` to `os.path.join(input_path, "..", "..", "input_files", etc.)`

The whole point of using the join function is to eliminate the `/` character.

Going even further, I like `pathlib` much better, but that's extra credit.

lamkina commented on tests/reg_tests/test_MPhysGeo.py:
```diff
@@ -0,0 +1,566 @@
+        self.prob.run_model()
+
+        totals = self.prob.check_totals(step=1e-7, out_stream=None)
+        assert_check_totals(totals)
```
Can we add hard-coded tolerances for this check as well? Better to have them here than use the defaults from OM incase they change in the future.

lamkina commented on tests/reg_tests/test_MPhysGeo.py:
```diff
@@ -0,0 +1,566 @@
+        paramKwargs.update(kwargs)
+
+        meshFile = os.path.join(input_path, "../../input_files/2x1x8_rectangle.stl")
+        ffdFile = os.path.join(input_path, "../../input_files/2x1x8_rectangle.xyz")
```
See path comment above.

lamkina commented on tests/reg_tests/test_MPhysGeo.py:
```diff
@@ -0,0 +1,566 @@
+        self.prob.run_model()
+
+        totals = self.prob.check_totals(step=1e-7, out_stream=None)
+        assert_check_totals(totals)
```
We should add tolerances here.

lamkina commented on pygeo/mphys/mphys_dvgeo.py:
```diff
@@ -201,7 +201,11 @@ def nom_addPointSet(self, points, ptName, add_output=True, DVGeoName=None, **kwa
-        DVGeo.addPointSet(points.reshape(len(points) // 3, 3), ptName, **kwargs)
+        if isinstance(DVGeo, DVGeometryESP):
+            # DVGeoESP can return a value to check the pointset distribution
+            dMax_global = DVGeo.addPointSet(points.reshape(len(points) // 3, 3), ptName, **kwargs)
```
Keep this as camel case: `dMaxGlobal`

lamkina reviewed (changes requested):
Thanks for the hard work @hajdik and @eytanadler . Just a few small changes, but overall this is great.

lamkina reviewed (dismissed):
Great stuff. Thanks!

marcomangano reviewed (approved):
Not very familiar with the MPhys wrapper here, but the code changes make sense and are well implemented. Good job!

marcomangano commented:
Is this change worth version update @eytanadler @hajdik ? Feel free to merge otherwise

eytanadler commented:
Yeah probably a good idea

lamkina reviewed (approved):
LGTM. @hajdik go ahead and merge this when you get the chance.
